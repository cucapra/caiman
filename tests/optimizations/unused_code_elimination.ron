// ARGS: -t
(
	version : (0, 0, 1),
	program : (
		types : {0:I32},
		funclets : {
			0 : (
                // empty funclet - should remain empty, not cause crashes, etc
                kind : MixedImplicit,
                input_types : [],
                output_types : [],
                nodes : [],
                tail_edge : Return(return_values : []) 
            ),
            1 : (
                // every node is used, nothing should be changed
                kind : MixedImplicit,
                input_types : [0],
                output_types : [0],
                nodes : [
                    Phi(index : 0),
                    CallExternalCpu(external_function_id : 0, arguments : [0]),
                    ExtractResult(node_id : 1, index : 0),
                ],
                tail_edge : Return(return_values : [2]) 
            ),
            2 : (
                // no nodes are used, they should all be pruned
                kind : MixedImplicit,
                input_types : [0],
                output_types : [],
                nodes : [
                    Phi(index : 0),
                    CallExternalCpu(external_function_id : 0, arguments : [0]),
                    ExtractResult(node_id : 1, index : 0),
                ],
                tail_edge : Return(return_values : []) 
            ),
            3 : (
                kind : MixedImplicit,
                input_types : [0],
                output_types : [0],
                nodes : [
                    Phi(index : 0),
                    CallExternalCpu(external_function_id : 0, arguments : []),
                    ExtractResult(node_id : 1, index : 0),
                    CallExternalCpu(external_function_id : 1, arguments : [2]),
                    ExtractResult(node_id : 3, index : 0),
                    CallExternalCpu(external_function_id : 2, arguments: [0]),
                    ExtractResult(node_id : 5, index : 0)
                ],
                tail_edge : Return(return_values : [4]) 
            ),
            4 : (
                kind : MixedImplicit,
                input_types : [0, 0, 0, 0],
                output_types : [0, 0],
                nodes : [
                    Phi(index : 0),
                    Phi(index : 1),
                    Phi(index : 2),
                    Phi(index : 3),

                    CallExternalCpu(external_function_id : 0, arguments : [0, 2]),
                    ExtractResult(node_id : 4, index : 0),

                    CallExternalCpu(external_function_id : 0, arguments : [1, 3]),
                    ExtractResult(node_id : 6, index : 0),

                    ConstantInteger(value : 42, type_id : 0),
                    CallExternalCpu(external_function_id : 0, arguments : [3, 8]),
                    ExtractResult(node_id : 9, index : 0),

                    CallExternalCpu(external_function_id : 0, arguments : [5, 8]),
                    ExtractResult(node_id : 11, index : 0)
                ],
                tail_edge : Return(return_values : [1, 12]) 
            )
        }
	)
)