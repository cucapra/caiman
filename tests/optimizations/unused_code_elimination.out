(
    version: (0, 0, 1),
    program: (
        types: {
            0: I32,
        },
        funclets: {
            0: (
                kind: MixedImplicit,
                input_types: [],
                output_types: [],
                nodes: [],
                tail_edge: Return(
                    return_values: [],
                ),
                input_resource_states: [],
                output_resource_states: [],
                local_meta_variables: {},
            ),
            1: (
                kind: MixedImplicit,
                input_types: [
                    0,
                ],
                output_types: [
                    0,
                ],
                nodes: [
                    Phi(
                        index: 0,
                    ),// [0]
                    CallExternalCpu(
                        external_function_id: 0,
                        arguments: [
                            0,
                        ],
                    ),// [1]
                    ExtractResult(
                        node_id: 1,
                        index: 0,
                    ),
                ],
                tail_edge: Return(
                    return_values: [
                        2,
                    ],
                ),
                input_resource_states: [],
                output_resource_states: [],
                local_meta_variables: {},
            ),
            2: (
                kind: MixedImplicit,
                input_types: [
                    0,
                ],
                output_types: [],
                nodes: [],
                tail_edge: Return(
                    return_values: [],
                ),
                input_resource_states: [],
                output_resource_states: [],
                local_meta_variables: {},
            ),
            3: (
                kind: MixedImplicit,
                input_types: [
                    0,
                ],
                output_types: [
                    0,
                ],
                nodes: [
                    CallExternalCpu(
                        external_function_id: 0,
                        arguments: [],
                    ),// [0]
                    ExtractResult(
                        node_id: 0,
                        index: 0,
                    ),// [1]
                    CallExternalCpu(
                        external_function_id: 1,
                        arguments: [
                            1,
                        ],
                    ),// [2]
                    ExtractResult(
                        node_id: 2,
                        index: 0,
                    ),
                ],
                tail_edge: Return(
                    return_values: [
                        3,
                    ],
                ),
                input_resource_states: [],
                output_resource_states: [],
                local_meta_variables: {},
            ),
            4: (
                kind: MixedImplicit,
                input_types: [
                    0,// [0]
                    0,// [1]
                    0,// [2]
                    0,
                ],
                output_types: [
                    0,// [0]
                    0,
                ],
                nodes: [
                    ConstantInteger(
                        value: 42,
                        type_id: 0,
                    ),// [0]
                    Phi(
                        index: 2,
                    ),// [1]
                    Phi(
                        index: 0,
                    ),// [2]
                    CallExternalCpu(
                        external_function_id: 0,
                        arguments: [
                            2,// [0]
                            1,
                        ],
                    ),// [3]
                    ExtractResult(
                        node_id: 3,
                        index: 0,
                    ),// [4]
                    CallExternalCpu(
                        external_function_id: 0,
                        arguments: [
                            4,// [0]
                            0,
                        ],
                    ),// [5]
                    ExtractResult(
                        node_id: 5,
                        index: 0,
                    ),// [6]
                    Phi(
                        index: 1,
                    ),
                ],
                tail_edge: Return(
                    return_values: [
                        7,// [0]
                        6,
                    ],
                ),
                input_resource_states: [],
                output_resource_states: [],
                local_meta_variables: {},
            ),
        },
        external_cpu_functions: [],
        external_gpu_functions: [],
        value_functions: {},
        pipelines: [],
    ),
)